Corrects the scancode for numlock, pause

ibeckermayer opened the pull request:
Adds capability for sending multiple scancodes for a single key in the process. The correct values were found in https://github.com/FreeRDP/FreeRDP/blob/ba8cf8cf2158018fb7abbedb51ab245f369be813/include/freerdp/scancode.h.

(This was discovered while investigating https://github.com/gravitational/teleport/issues/38571, I don't know whether it might solve that issue since I wasn't able to reproduce it, however I doubt it).

changelog: fixes the NumLock and Pause keys for Desktop Access sessions

Commit:
Corrects the scancode for numlock, adds the capability of sending multiple scancodes for a single keypress in cases such as pause

zmb3 reviewed (commented):
What about backwards compatibility? Do we need a new message ID for this?

zmb3 commented on web/packages/teleport/src/lib/tdp/codec.ts:
```diff
@@ -523,19 +529,38 @@ export default class Codec {
-    const buffer = new ArrayBuffer(6);
+
+    // Normalize scanCodeOrCodes to always be an array
+    const scanCodes = Array.isArray(scanCodeOrCodes)
```
Why not just fix the map above once and then avoid doing this work at runtime?

zmb3 commented on web/packages/teleport/src/lib/tdp/codec.ts:
```diff
@@ -523,19 +529,38 @@ export default class Codec {
+
   // encodeKeyboardInput encodes a keyboard action.
   // Returns null if an unsupported code is passed.
   // | message type (5) | key_code uint32 | state byte |
```
This is no longer accurate, right?

Which would mean we need to update the RFD and the backend as well. (And probably use a new message ID)

zmb3 commented on web/packages/teleport/src/lib/tdp/codec.ts:
```diff
@@ -523,19 +529,38 @@ export default class Codec {
-    view.setUint8(5, state);
+
+    let offset = 0;
+    scanCodes.forEach(scanCode => {
```
Now I see what you are doing here. You are actually sending multiple TDP messages in a single websocket write. We've never done that before, it's always been 1:1. 

I'm not sure there's a compelling enough reason to break from tradition here.

ibeckermayer replied:
Hmm good point, we do have some logic on the backend that may rely on this 1:1. I'll fix that.

gzdunek commented on web/packages/teleport/src/lib/tdp/codec.ts:
```diff
@@ -523,19 +529,38 @@ export default class Codec {
     return buffer;
   }
 
+  static keyboardButtonMessageLength =
```
Is there any reason to define this as a class field?
I think the more common practice is to put it as a constant on top of the file.
```ts
const KEYBOARD_BUTTON_MESSAGE_LENGTH = ...
```

BTW I would do the same for `_keyScancodes`, it is a readonly value.

Commit:
Code Review:

Moves _keyScancodes into a readonly const, ensures that we only ever
send a single tdp message per websocket message.

zmb3 commented on web/packages/teleport/src/lib/tdp/client.ts:
```diff
@@ -611,9 +611,12 @@ export default class Client extends EventEmitterWebAuthnSender {
-    const msg = this.codec.encodeKeyboardInput(code, state);
-    if (msg) this.send(msg);
+    const msgs = this.codec.encodeKeyboardInput(code, state);
+    if (msgs) {
```
If you make `encodeKeyboardInput` always return an array (which can be empty) then the typing becomes simpler and this `if` can go away. (`forEach` on an empty array is a no-op)

ibeckermayer replied:
adf8c2fbf9eb4d7a2813aa4d41d75048ae7e18f3

gzdunek commented on web/packages/teleport/src/lib/tdp/codec.ts:
```diff
@@ -526,15 +369,25 @@ export default class Codec {
+    );
+  }
+
+  private encodeKeyboardInputSingle(
```
?
```suggestion
  private encodeKeyScancode(
```

ibeckermayer replied:
adf8c2fbf9eb4d7a2813aa4d41d75048ae7e18f3

gzdunek reviewed (approved):
Thanks for extracting constants :)

Commit:
CR
